fflib/src/classes/fflib_SObjectUnitOfWork.cls

```diff
@@ -667,6 +747,15 @@ public virtual class fflib_SObjectUnitOfWork
 		}
 	}
 
+    private void upsertDmlByType() 
+    {
+        String sobjName;
+        for (Schema.SObjectType sObjectType : m_sObjectTypes) {
+            sobjName = sObjectType.getDescribe().getName();
+			m_dml.dmlUpsert(m_upsertRecordsPerType.get(sobjName), m_externalIdToUpsertPerType.get(sobjName));
+		}
```

Comment:
I think you are missing here the resolving of relationships. You do register them in the registerupsert method

Reply:
Seems like it, I based my update in the `registetDirty` method that seems to also miss the resolving, is it a bug in the dirty relationship resolution or are the dirty relationships resolved elsewhere?

Reply:
Resolving of relationships is only for new records, but since an upsert also handles new records we should resolve the relationships here.

Reply:
Working on that I noted the following:

Since the `insertDmlByType` method iterates over all sObjects for the Unit of Work (not necessarily the records that were registered with `registerNew`) the `insertDmlByType` will always resolve all relationships (even those registered using one of the `registerDirty` methods).

Adding another resolver in the `upsertDmlByType` method will result in resolving again the relationships which could lead to some errors like unable to set again `creatable` but not `updateable` fields (such as the `Product2Id` in a `PricebookEntry`).

That's why I need to modify the `resolve` method on the `Relationship` class, in order to resolve whats missing resolution. What are your thoughts about it?

Reply:
Looking again, that change I made on the resolver is not a good one since it will ignore initial resolvers if the field is already filled before registering and therefore producing unexpected results. 

An idea comes to my head that maybe each `xByType` method should be responsible to resolve the relationships related only for the records registered for that operation, although the risk exists that a greater change in the implementation may be needed

Reply:
In my opinion seems like the way forward with less impact would be to leave the only one resolver in the insertDmlByType method, since it will be called always even when there's no records to insert. If you confirm I rollback that commit

Reply:
If you agree with this idea I can work on the necessary changes

Reply:
I didn't notice the thumbs up from the team on this old pending contribution, I will try to work on it again and update it. So sorry for the abandoned pull request

Reply:
@jonathanwiesel If you want any help on this, please let me know. My current implementation is needing this functionality and I was about to resurrect this PR for my own purposes, but happy to help you instead. 👍

Reply:
Thank you for the support @ctchipps 👍  I think I got it, I'll let you know in in case it get's nasty

Reply:
Ok so what I ended up doing to maintain most of the logic intact to avoid refactoring was in fact adding that second resolver to the upsert method, but adding a conditional in the resolver logic to check if a relationship has already been resolved to not resolve it again (and therefore mitigate risks of double resolving on `creatable` but no `updatable` fields).

What do you guys think?

Reply:
@jonathanwiesel Awesome work! I am going to try to get to reviewing this sometime this week.

@wimvelzeboer Do you mind to take a review pass on this as well?

Reply:
@ctchipps @wimvelzeboer I was wondering if you had the opportunity to take a look at this?